src/php/parser/Parser.php

```diff
@@ -0,0 +1,530 @@
+                }
+            }
+        }
+
+        throw new RuntimeException("must not be here!");
+    }
+
+    private function moveEval(EvalAction $action): Action
+    {
+        $expression = $action->getExpression();
+        $position = $action->getPosition();
+        $asts = $action->getAsts();
+        $error = $action->getError();
+        $continuations = $action->getContinuations();
+        $eof = $position >= strlen($this->input);
+
+        switch ($expression->getType()) {
+            case ExpressionType::ANY_CHAR:
+            {
+                if ($eof) {
+                    $matchResult = $this->reject($this->updateError($error, $position, new WildcardError()));
+                } else {
+                    $expected = mb_substr($this->input, $position, 1);
+                    $ord = IntlChar::ord($expected);
+
+                    $matchResult = $this->accept($position + strlen($expected), ASTs::asts(new Char($ord, $position), $asts), $error);
+                }
+
+                return $this->applyNext($continuations, $matchResult);
+            }
+            case ExpressionType::ALT:
+            {
+                $expressions = Expression::asAltExpression($expression)->getExpressions();
+
+                if ($expressions->isEmpty()) {
+                    return $this->applyNext($continuations, $this->reject($error));
+                } else {
+                    $cons = Continuations::cons(new AltContinuation($expressions->tail(), $position, $asts), $continuations);
+                    return $this->evalNext($expressions->head(), $position, $asts, $error, $cons);
+                }
+            }
+            case ExpressionType:: AND:
+            {
+                $cons = Continuations::cons(new AndContinuation($position, $asts, $error), $continuations);
+                return $this->evalNext($expression->getExpression(), $position, new ASTs(), $error, $cons);
+            }
+            case ExpressionType::NOT:
+            {
+                $cons = Continuations::cons(new NotContinuation($position, $asts, $error), $continuations);
+                return $this->evalNext($expression->getExpression(), $position, new ASTs(), $error, $cons);
+            }
+            case ExpressionType::VOID:
+            {
+                $cons = Continuations::cons(new VoidContinuation($asts), $continuations);
+                return $this->evalNext($expression->getExpression(), $position, new ASTs(), $error, $cons);
+            }
+            case ExpressionType::CHAR:
+            {
+                $expression = CharExpression::asCharExpression($expression);
+
+                if ($eof) {
+                    $matchResult = $this->reject($this->updateError($error, $position, new CharacterError($expression)));
+                } else {
+                    $expected = Expression::asCharExpression($expression)->getChar();
+                    $char = mb_substr($this->input, $position, 1);
```

Comment:
`mb_substr` is `O(N)` for UTF-8 according to this: https://www.php.net/manual/en/function.mb-substr.php#117764